presto-memsql/src/main/java/io/prestosql/plugin/memsql/MemSqlClient.java

```diff
@@ -88,6 +132,118 @@ protected String getTableSchemaName(ResultSet resultSet)
         return resultSet.getString("TABLE_CAT");
     }
 
+    @Override
+    public Optional<ColumnMapping> toPrestoType(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        String jdbcTypeName = typeHandle.getJdbcTypeName()
+                .orElseThrow(() -> new PrestoException(JDBC_ERROR, "Type name is missing: " + typeHandle));
+
+        Optional<ColumnMapping> mapping = getForcedMappingToVarchar(typeHandle);
+        if (mapping.isPresent()) {
+            return mapping;
+        }
+        Optional<ColumnMapping> unsignedMapping = getUnsignedMapping(typeHandle);
+        if (unsignedMapping.isPresent()) {
+            return unsignedMapping;
+        }
+
+        if (jdbcTypeName.equalsIgnoreCase("json")) {
+            return Optional.of(jsonColumnMapping());
+        }
+
+        int columnSize = typeHandle.getColumnSize();
+        switch (typeHandle.getJdbcType()) {
+            // TODO not all these type constants are necessarily used by the JDBC driver
+            case Types.VARCHAR:
+            case Types.NVARCHAR:
+            case Types.LONGVARCHAR:
+            case Types.LONGNVARCHAR:
+                VarcharType varcharType = (columnSize <= VarcharType.MAX_LENGTH) ? createVarcharType(columnSize) : createUnboundedVarcharType();
+                // Remote database can be case insensitive.
+                PredicatePushdownController predicatePushdownController = PUSHDOWN_AND_KEEP;
+                return Optional.of(ColumnMapping.sliceMapping(varcharType, varcharReadFunction(varcharType), varcharWriteFunction(), predicatePushdownController));
+            // TODO handle decimals if decimal_rounding_mode session property is needed
+        }
+
+        // TODO add explicit mappings
+        return super.toPrestoType(session, connection, typeHandle);
+    }
+
+    @Override
+    public WriteMapping toWriteMapping(ConnectorSession session, Type type)
+    {
+        if (REAL.equals(type)) {
+            return WriteMapping.longMapping("float", realWriteFunction());
+        }
+        // TODO implement TIME types
+        // TODO add support for other TIMESTAMP precisions
+        if (TIMESTAMP_MILLIS.equals(type)) {
+            return WriteMapping.longMapping("datetime", timestampWriteFunction(TIMESTAMP_MILLIS));
+        }
+        if (VARBINARY.equals(type)) {
+            return WriteMapping.sliceMapping("mediumblob", varbinaryWriteFunction());
+        }
+        if (type instanceof VarcharType) {
+            VarcharType varcharType = (VarcharType) type;
+            String dataType;
+            if (varcharType.isUnbounded()) {
+                dataType = "longtext";
+            }
+            else if (varcharType.getBoundedLength() <= 255) {
```

Comment:
We can replace this with

```suggestion
            else if (varcharType.getBoundedLength() <= 21845) {
                dataType = format("varchar(%s)", varcharType.getBoundedLength());
            }
```

This will allow the varchars in TPCH to look correct (ie. varchar(1) etc.). I kept it this way because MySQL has a similar implementation even though we can do the same thing there.

Reply:
Mysql thing seems like legacy, i dont see a reason to follow it here

define a constant like MEMSQL_VARCHAR_MAX_LENGTH

Reply:
Fixed.